File: products/FG1-flowguard/tools/fg1_production_tester/lib/providers/bench_config.dart
import 'package:shared_preferences/shared_preferences.dart';

/// Bench-station settings -- set once per bench, persisted locally.
/// Mirrors the constants at the top of testing/test_production.py.
class BenchConfig {
  final String flashBridgeHost; // e.g. "192.168.1.50:8787"
  final String jigHost; // e.g. "fg1jig.local"
  final String firmwareEnv; // "esp32dev" or "esp32dev_ds1307"
  final String testApSsid; // Bench LAN, see TEST_JIG_SPEC.md section 2.5
  final String testApPassword;
  final int expectedCalibrationPpl;
  final int flowTestPulseCount;
  final String operatorName;
  final String stationName;

  const BenchConfig({
    this.flashBridgeHost = '',
    // Static IP the jig configures itself once it joins the DUT's
    // SoftAP -- not fg1jig.local, since stock Android's plain HTTP
    // client doesn't resolve mDNS hostnames (see jig firmware's
    // header comment). mDNS is still up for laptop-side debugging.
    this.jigHost = '192.168.4.50',
    this.firmwareEnv = 'esp32dev',
    this.testApSsid = 'FG1-TEST-STATION',
    this.testApPassword = '',
    this.expectedCalibrationPpl = 450,
    this.flowTestPulseCount = 450,
    this.operatorName = '',
    this.stationName = 'bench-1',
  });

  BenchConfig copyWith({
    String? flashBridgeHost,
    String? jigHost,
    String? firmwareEnv,
    String? testApSsid,
    String? testApPassword,
    int? expectedCalibrationPpl,
    int? flowTestPulseCount,
    String? operatorName,
    String? stationName,
  }) =>
      BenchConfig(
        flashBridgeHost: flashBridgeHost ?? this.flashBridgeHost,
        jigHost: jigHost ?? this.jigHost,
        firmwareEnv: firmwareEnv ?? this.firmwareEnv,
        testApSsid: testApSsid ?? this.testApSsid,
        testApPassword: testApPassword ?? this.testApPassword,
        expectedCalibrationPpl: expectedCalibrationPpl ?? this.expectedCalibrationPpl,
        flowTestPulseCount: flowTestPulseCount ?? this.flowTestPulseCount,
        operatorName: operatorName ?? this.operatorName,
        stationName: stationName ?? this.stationName,
      );

  static const _prefsPrefix = 'fg1_bench_config_';

  Future<void> save() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('${_prefsPrefix}flash_bridge_host', flashBridgeHost);
    await prefs.setString('${_prefsPrefix}jig_host', jigHost);
    await prefs.setString('${_prefsPrefix}firmware_env', firmwareEnv);
    await prefs.setString('${_prefsPrefix}test_ap_ssid', testApSsid);
    await prefs.setString('${_prefsPrefix}test_ap_password', testApPassword);
    await prefs.setInt('${_prefsPrefix}expected_ppl', expectedCalibrationPpl);
    await prefs.setInt('${_prefsPrefix}flow_pulse_count', flowTestPulseCount);
    await prefs.setString('${_prefsPrefix}operator', operatorName);
    await prefs.setString('${_prefsPrefix}station', stationName);
  }

  static Future<BenchConfig> load() async {
    final prefs = await SharedPreferences.getInstance();
    const d = BenchConfig();
    return BenchConfig(
      flashBridgeHost: prefs.getString('${_prefsPrefix}flash_bridge_host') ?? d.flashBridgeHost,
      jigHost: prefs.getString('${_prefsPrefix}jig_host') ?? d.jigHost,
      firmwareEnv: prefs.getString('${_prefsPrefix}firmware_env') ?? d.firmwareEnv,
      testApSsid: prefs.getString('${_prefsPrefix}test_ap_ssid') ?? d.testApSsid,
      testApPassword: prefs.getString('${_prefsPrefix}test_ap_password') ?? d.testApPassword,
      expectedCalibrationPpl: prefs.getInt('${_prefsPrefix}expected_ppl') ?? d.expectedCalibrationPpl,
      flowTestPulseCount: prefs.getInt('${_prefsPrefix}flow_pulse_count') ?? d.flowTestPulseCount,
      operatorName: prefs.getString('${_prefsPrefix}operator') ?? d.operatorName,
      stationName: prefs.getString('${_prefsPrefix}station') ?? d.stationName,
    );
  }
}

File: projects/tools/flasher-tester-app/app/src/main/java/com/nbagri/flashertester/BenchConfig.kt
package com.nbagri.flashertester

import android.content.Context

/**
 * Bench-station settings, persisted in SharedPreferences. Slimmer than
 * the laptop-based design (fg1_production_tester / this app's earlier
 * revision) -- no Flash Bridge host, no manual firmware-env string:
 * the firmware build comes from ApiClient's product/build picker, and
 * the jig is found automatically by probing the USB hub (see
 * MainActivity.identifyUsbPorts()), not configured by hand.
 */
data class BenchConfig(
    // "Test AP" the DUT's STA side joins for the WiFi+MQTT check --
    // the site's office WiFi router (see
    // docs/testing/PRODUCTION_TOOL_SPEC_V2.md §4.1). A phone-hotspot
    // test AP was tried and dropped 2026-09-16: hotspot-ON +
    // WiFi-client-joined-to-the-DUT's-SoftAP at the same time made
    // Android route the hotspot's own internet through that dead-end
    // WiFi link instead of cellular, breaking DNS for the DUT. The
    // office router has its own internet backhaul, so this whole class
    // of problem doesn't apply -- but it does mean these need to be set
    // explicitly per site in Settings before the first run; left blank
    // here deliberately rather than defaulting to stale hotspot
    // credentials that would silently point at the wrong network.
    val testApSsid: String = "",
    val testApPassword: String = "",
    val expectedCalibrationPpl: Int = 450,
    val flowTestPulseCount: Int = 450,
    val operatorName: String = "",
    val stationName: String = "bench-1",
) {
    fun save(context: Context) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit().apply {
            putString("test_ap_ssid", testApSsid)
            putString("test_ap_password", testApPassword)
            putInt("expected_ppl", expectedCalibrationPpl)
            putInt("flow_pulse_count", flowTestPulseCount)
            putString("operator", operatorName)
            putString("station", stationName)
            apply()
        }
    }

    companion object {
        private const val PREFS = "bench_config"

        /** Last device ID seen (from a boot log parse) -- separate from the
         *  rest of this class since it changes per-unit, not per-bench.
         *  Persisted (not just an in-memory field) so "Run Tests Only"
         *  still pre-fills correctly after an app restart -- an in-memory
         *  field alone left the operator guessing whether to type
         *  "SWC_001_B468" or "B468" every time. */
        fun saveLastDeviceId(context: Context, deviceId: String) {
            context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
                .putString("last_device_id", deviceId).apply()
        }

        fun loadLastDeviceId(context: Context): String? =
            context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getString("last_device_id", null)

        fun load(context: Context): BenchConfig {
            val p = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            val d = BenchConfig()
            return BenchConfig(
                testApSsid = p.getString("test_ap_ssid", d.testApSsid)!!,
                testApPassword = p.getString("test_ap_password", d.testApPassword)!!,
                expectedCalibrationPpl = p.getInt("expected_ppl", d.expectedCalibrationPpl),
                flowTestPulseCount = p.getInt("flow_pulse_count", d.flowTestPulseCount),
                operatorName = p.getString("operator", d.operatorName)!!,
                stationName = p.getString("station", d.stationName)!!,
            )
        }
    }
}
